docs(core-concepts): adds descriptions for greenhouse core concepts (#651)
cloudoperators/greenhouse: docs/getting-started/core-concepts/plugins.md

---
title: "PluginDefinitions and Plugins"
weight: 4
---

> 

Reviewer: ```suggestion
## Extending Greenhouse
Plugins are the Greenhouse way to extend core functionality with domain specific features. The plugin functionality in Greenhouse is provided via two resources: `PluginDefinitions` and `Plugins`.
```

> ## What are Plugin Definitions?

Reviewer: ```suggestion
## PluginDefinitions and Plugins
```

> PluginDefinitions are a way to extend the functionality of the platform. They can include a [Juno microfrontend](https://github.com/cloudoperators/juno) into the Greenhouse Dashboard and/or a backend component. The backend components of a PluginDefinition is packaged as a Helm Chart and provides default values for the Helm Chart. This allows to package and distribute tools such as Prometheus with a sensible default configuration, as well as giving the user a list of configurable values.

Reviewer: ```suggestion
`PluginDefinitions` are, as the name suggests the definition of a plugin, whereas a `Plugin` is an instance of said `PluginDefinition`. There can be only one `PluginDefinition` of a certain `name`, whereas there can be multiple `Plugins` (also of a unique `name`) referring to a single `PluginDefinition`.

`PluginDefinitions` are singletons across one Greenhouse installation (cluster scoped resource) whereas `Plugins` live within an Organization (namespaced resources).

A `PluginDefinition` can include a frontend into the Greenhouse Dashboard and/or a backend component. The frontend is expected to be a standalone microfrontend. Greenhouse encourages the use of the [Juno framework](https://github.com/cloudoperators/juno) to create microfrontends.
The backend components of a `PluginDefinition` is packaged as a Helm Chart and providing sane and opinionated default values. This allows us to package and distribute tools such as Prometheus with a sensible default configuration, as well as giving the user a list of configurable values.
```

> ## Plugins

Reviewer: ```suggestion
```

> In order to deploy the Helm Chart of a Plugin Definition into a particular cluster a Plugin is used. The Plugin specifies the PluginDefintion to deploy, the release namespace to use, the values to set and the cluster to deploy to. Depending on the PluginDefinition it may be required to provide required values (e.g. credentials, endpoints etc.), but in general the PluginDefinition provides sensible defaults.

Reviewer: ```suggestion
In order to actually run an instance of a `PluginDefinition` a `Plugin` is used. The `Plugin` specifies which concrete values to apply to a `PluginDefinition`, which `Cluster` to run the backend component on and what namespace it should be deployed to.
```

> In order to deploy the Helm Chart of a Plugin Definition into a particular cluster a Plugin is used. The Plugin specifies the PluginDefintion to deploy, the release namespace to use, the values to set and the cluster to deploy to. Depending on the PluginDefinition it may be required to provide required values (e.g. credentials, endpoints etc.), but in general the PluginDefinition provides sensible defaults.

Reviewer 2: ```suggestion
A Plugin is used to deploy a Helm release described in a PluginDefinition. The Plugin specifies the PluginDefinition to be installed, the option values to be set as well as the cluster and namespace to be deployed to. Depending on the PluginDefinition, it may be necessary to specify required values (e.g. credentials, endpoints, etc.), but in general the PluginDefinition provides well-established default values.
```

> [!NOTE]
> In this example the Plugin 'openTelemetry-cluster-a' is used to deploy the PluginDefinition 'openTelemetry' to the cluster 'cluster-a'.

```mermaid
flowchart TB
  subgraph "Greenhouse" 

>     subgraph "Organization" 

Reviewer: ```suggestion
    subgraph "namespace: <org-name>" 
```

>         pluginDefinition[PluginDefinition 'openTelemetry']

Reviewer: This should probably live outside the org's namespace?

        clusterA[Cluster 'cluster-a']
        plugin1[Plugin 'openTelemetry-cluster-a']
    end
  end

  subgraph "cluster-a"
    subgraph "observability"
      plugin1HelmRelease[HelmRelease 'openTelemetry-cluster-a']
    end
  end

  plugin1 -.- clusterA
  plugin1 -.- pluginDefinition
  plugin1 --"deploys"--> plugin1HelmRelease
```

## PluginPresets

PluginPresets are a mechanism to configure Plugins for multiple clusters at once. They are used to define a common configuration for a PluginDefinition that can be applied to multiple clusters, while allowing to override the configuration for individual clusters.

> [!NOTE]
> In this example the PluginPreset 'example-obs' references the PluginDefinition 'example' and contains a clusterSelector that matches the clusters 'cluster-a' and 'cluster-b'. The PluginPreset creates two Plugins 'example-obs-cluster-a' and 'example-obs-cluster-b' for the respective clusters.

```mermaid
flowchart TB
  subgraph "Greenhouse" 

>     subgraph "Organization" 

Reviewer: ```suggestion
    subgraph "namespace: <org-name>" 
```

>         pluginDefinition[PluginDefinition 'example']

Reviewer: move out of namespace `<org-name>`?

---
title: "PluginDefinitions and Plugins"
weight: 4
---

## What are PluginDefinitions and Plugins?

PluginDefinitons and Plugins are the Greenhouse way to extend the core functionality with domain specific features.
PluginDefinitions, as the name suggests, are the definition of a Plugin, whereas a Plugin is a concrete instance of a PluginDefinition that is deployed to a Cluster.

The PluginDefinitions are shared between all Organizations in Greenhouse. A PluginDefinition can include a frontend, that is displayed in the Greenhouse dashboard and/or a backend component. The frontend is expected to be a standalone microfrontend created with the [Juno framework](https://github.com/cloudoperators/juno).
The backend components of a PluginDefinition are packaged as a Helm Chart and provide sane and opinionated default values. This allows Greenhouse to package and distribute tools such as Prometheus with a sensible default configuration, as well as giving the user a list of configurable values.

> A Plugin is used to deploy the Helm Chart referenced referenced in a PluginDefinition to a Cluster. The Plugin can be considered as an instance of a PluginDefinition, this instance specifies the PluginDefinition, the desired Cluster and additional values to set. Depending on the PluginDefinition, it may be necessary to specify required values (e.g. credentials, endpoints, etc.), but in general the PluginDefinition provides well-established default values.

Reviewer 2: ```suggestion
A Plugin is used to deploy the Helm Chart referenced in a PluginDefinition to a Cluster. The Plugin can be considered as an instance of a PluginDefinition, this instance specifies the PluginDefinition, the desired Cluster and additional values to set. Depending on the PluginDefinition, it may be necessary to specify required values (e.g. credentials, endpoints, etc.), but in general the PluginDefinition provides well-established default values.
```
